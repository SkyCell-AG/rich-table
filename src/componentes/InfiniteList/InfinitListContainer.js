import React, {
    useCallback,
    useEffect,
    useRef,
    forwardRef,
    useLayoutEffect,
    useReducer,
    useMemo,
} from 'react'
import noop from 'lodash/noop'
import get from 'lodash/get'
import PropTypes from 'prop-types'

import * as statuses from 'utils/requestStatuses'
import createCancelablePromise from 'utils/createCancelablePromise'

import {
    loadDataSuccess,
    loadDataFailure,
    loadDataPending,
    updateDataEntry,
    addDataEntry,
} from './store/actions'
import reducer from './store/reducer'
import InfiniteList from './InfiniteList'

const propTypes = {
    Row: PropTypes.func.isRequired,
    uniqField: PropTypes.string,
    load: PropTypes.func.isRequired,
    onUpdateMatchedResults: PropTypes.func,
}

const defaultProps = {
    onUpdateMatchedResults: noop,
    uniqField: 'id',
}

const InfiniteListContainer = forwardRef((props, ref) => {
    const {
        load,
        onUpdateMatchedResults,
        uniqField,
        Row,
    } = props

    const [
        {
            data,
            matchedResults,
            status,
            page,
        },
        dispatch,
    ] = useReducer(reducer, {
        status: statuses.PRISTIN,
        data: [],
        matchedResults: 0,
        page: 0,
    })

    useEffect(() => {
        ref.current = { // eslint-disable-line
            update: (updatedElement) => {
                dispatch(updateDataEntry(updatedElement, uniqField))
            },
            add: (updatedElement) => {
                dispatch(addDataEntry(updatedElement))
            },
        }
    }, [
        ref,
        uniqField,
    ])

    const wrapperRef = useRef(null)
    const spacerRef = useRef(null)
    const cancelRequest = useRef(null)

    const hasMore = useMemo(() => {
        return data.length < matchedResults
    }, [
        data.length,
        matchedResults,
    ])

    const loadNewPage = useCallback((oldVal, newPage) => {
        dispatch(loadDataPending())

        const [
            request,
            cancel,
        ] = createCancelablePromise(load(newPage))

        cancelRequest.current = cancel

        request
            .then((response) => {
                if (!get(response, 'data') || get(response, 'meta.matchedresults') === undefined) {
                    throw new Error('Not valid response')
                }

                return response
            })
            .then(({
                data: newPageData,
                meta: {
                    matchedresults,
                },
            }) => {
                cancelRequest.current = null
                dispatch(loadDataSuccess({
                    meta: {
                        page: newPage,
                        matchedResults: matchedresults,
                    },
                    data: [
                        ...oldVal,
                        ...newPageData,
                    ],
                }))
            })
            .catch((err) => {
                if (err.message === 'canceled') {
                    return
                }

                dispatch(loadDataFailure(err))
            })
    }, [
        load,
        dispatch,
    ])

    const loadNextPage = useCallback(() => {
        loadNewPage(data, page + 1)
    }, [
        loadNewPage,
        data,
        page,
    ])

    const onScroll = useCallback(({
        target,
    }) => {
        if (
            status === statuses.PENDING
            || !hasMore
            || target !== wrapperRef.current
            || target.scrollHeight - target.scrollTop > target.offsetHeight * 1.2
        ) {
            return
        }

        loadNextPage()
    }, [
        status,
        hasMore,
        loadNextPage,
    ])

    useEffect(() => {
        if (cancelRequest.current) {
            cancelRequest.current()
        }

        loadNewPage([], 1)
    }, [loadNewPage])

    useEffect(() => {
        onUpdateMatchedResults(matchedResults)
    }, [
        onUpdateMatchedResults,
        matchedResults,
    ])

    useLayoutEffect(() => {
        if (status !== statuses.SUCCESS) {
            return
        }

        const wrapper = wrapperRef.current
        const spacer = spacerRef.current

        if (!hasMore) {
            spacer.style.height = '0px'
            return
        }

        const {
            height: wrapperHeight,
        } = wrapper.getBoundingClientRect()

        const children = [...wrapper.childNodes]
        const childrenHeight = children.reduce((sum, element) => {
            if (element === spacer) {
                return sum
            }

            const {
                height,
            } = element.getBoundingClientRect()

            return sum + height
        }, 0)

        const heightDiff = wrapperHeight - childrenHeight

        if (heightDiff > 0) {
            spacer.style.height = `${heightDiff + 20}px`
        } else {
            spacer.style.height = '0px'
        }
    }, [
        data,
        status,
        hasMore,
    ])

    return (
        <InfiniteList
            {...props}
            Row={Row}
            wrapperRef={wrapperRef}
            spacerRef={spacerRef}
            hasMore={hasMore}
            data={data}
            status={status}
            loadNextPage={loadNextPage}
            onScroll={status === statuses.SUCCESS || statuses.PRISTIN ? onScroll : noop}
        />
    )
})

InfiniteListContainer.propTypes = propTypes
InfiniteListContainer.defaultProps = defaultProps

export default InfiniteListContainer
