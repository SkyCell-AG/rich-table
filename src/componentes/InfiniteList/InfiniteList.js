import React from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import CircularProgress from '@mui/material/CircularProgress'
// import SnackbarContent from '@mui/material/SnackbarContent'

import {
    PENDING,
    FAILURE,
    SUCCESS,
    requestType,
} from 'utils/requestStatuses'

import useStyles from './InfiniteList.style'

const propTypes = {
    data: PropTypes.array, // eslint-disable-line
    status: requestType.isRequired,
    onScroll: PropTypes.func.isRequired,
    className: PropTypes.string,
    Row: PropTypes.func.isRequired,
    BeforeList: PropTypes.func,
}

const defaultProps = {
    data: [],
    className: '',
    BeforeList: null,
}

const InfiniteList = ({
    data,
    status,
    onScroll,
    wrapperRef, // eslint-disable-line
    spacerRef, // eslint-disable-line
    className,
    BeforeList,
    Row,
}) => {
    const classes = useStyles()

    return (
        <div
            onScroll={onScroll}
            ref={wrapperRef}
            className={clsx(classes.wrapper, className)}
            data-testid="infinitelist"
        >
            {BeforeList && <BeforeList data={data} />}
            {status === FAILURE && (
                <div className={classes.noDataContainer}>
                    <div className={classes.noData}>
                        Something went wrong and the data could not be loaded.
                    </div>
                </div>
            )}
            {SUCCESS === status && data.length === 0 && (
                <div className={classes.noDataContainer}>
                    <div className={classes.noData}>
                        There is no data in this table yet.
                    </div>
                </div>
            )}
            {data.map(Row)}
            {status === PENDING && (
                <CircularProgress
                    className={classes.loader}
                    size={40}
                />
            )}
            <div ref={spacerRef} />
        </div>
    )
}

InfiniteList.propTypes = propTypes
InfiniteList.defaultProps = defaultProps

export default InfiniteList
