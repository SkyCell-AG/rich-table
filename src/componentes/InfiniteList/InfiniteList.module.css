.wrapper {
    height: 100%;
    overflow: auto;
    background-color: #FFFFFF;
    border-radius: 10px 10px 0px 0px;
    box-shadow: 0px 1px 3px #00000033;
}

.loadMoreContainer {
    display: flex;
    justify-content: center;
}

.loadMore {
    position: absolute;
    margin-top: -64px;
}

.loader {
    position: absolute;
    left: 50%;
    top: 50%;
    margin-left: -20px;
    margin-top: -20px;
}

.failureMessage {
    z-index: 2;
    position: sticky;
    top: 0;
    background-color: #d32f2f;
}
