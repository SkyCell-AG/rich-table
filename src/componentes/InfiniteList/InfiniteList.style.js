import {
    makeStyles,
} from '@mui/styles'
import {
    useTheme,
} from '@mui/material/styles'

const useStyles = makeStyles(() => {
    const theme = useTheme()

    return {
        wrapper: {
            height: '100%',
            overflow: 'auto',
            backgroundColor: theme.palette.common.white,
            borderRadius: '10px 10px 0px 0px',
            boxShadow: '0px 1px 3px #00000033',
        },
        loadMoreContainer: {
            display: 'flex',
            justifyContent: 'center',
        },
        loadMore: {
            position: 'absolute',
            marginTop: '-64px',
        },
        loader: {
            position: 'absolute',
            left: '50%',
            top: '80%',
            marginLeft: '-20px',
            marginTop: '-20px',
            zIndex: 2,
        },
        failureMessage: {
            zIndex: 2,
            position: 'sticky',
            top: 0,
            backgroundColor: '#d32f2f',
        },
        pendingContainer: {
            marginTop: theme.spacing(6.875),
        },
        noDataContainer: {
            display: 'flex',
            justifyContent: 'center',
            marginTop: theme.spacing(3),
        },
        noData: {
            fontSize: theme.typography.body1.fontSize,
            color: theme.typography.body1.color,
            fontWeight: theme.typography.body1.fontWeight,
        },
    }
})

export default useStyles
