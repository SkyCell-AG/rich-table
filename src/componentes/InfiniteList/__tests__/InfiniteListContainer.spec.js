import React from 'react'
import {
    shallow,
} from 'enzyme'
import delay from 'lodash/delay'

import InfiniteListContainer from '../InfinitListContainer'
import InfiniteList from '../InfiniteList'

describe('InfinitListContainer', () => {
    test('snapshot on load Success', () => {
        const load = jest.fn().mockImplementation((d) => {
            return new Promise((resolve) => {
                setTimeout(() => {
                    return resolve({
                        meta: {
                            matchedresults: 20,
                        },
                        data: [{
                            val: d,
                        }],
                    })
                }, 200)
            })
        })

        const infiniteListContainerElement = shallow(
            <InfiniteListContainer
                Row={jest.fn()}
                uniqField="id"
                load={load}
                onUpdateMatchedResults={jest.fn()}
                renderFailureMessage={jest.fn()}
                renderEmptyMessage={jest.fn()}
            />,
        )

        delay(() => {
            infiniteListContainerElement.find(InfiniteList).props().loadNextPage()
            infiniteListContainerElement.find(InfiniteList).props().loadNextPage()
        }, 100)
        delay(() => {
            infiniteListContainerElement.find(InfiniteList).props().loadNextPage()
        }, 110)
        delay(() => {
            infiniteListContainerElement.find(InfiniteList).props().loadNextPage()
        }, 300)

        return new Promise((resolve) => {
            setTimeout(resolve, 400)
        })
            .then(() => {
                infiniteListContainerElement.find(InfiniteList).props().loadNextPage()

                return new Promise((resolve2) => {
                    setTimeout(resolve2, 500)
                })
            })
            .then(() => {
                expect(
                    infiniteListContainerElement,
                ).toMatchSnapshot()
            })
    })
})
