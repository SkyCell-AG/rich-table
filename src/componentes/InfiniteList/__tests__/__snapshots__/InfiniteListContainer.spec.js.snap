// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfinitListContainer snapshot on load Success 1`] = `
<InfiniteList
  BeforeList={null}
  Row={[MockFunction]}
  className=""
  data={
    Array [
      Object {
        "val": 1,
      },
      Object {
        "val": 2,
      },
    ]
  }
  hasMore={true}
  load={
    [MockFunction] {
      "calls": Array [
        Array [
          1,
        ],
        Array [
          1,
        ],
        Array [
          1,
        ],
        Array [
          1,
        ],
        Array [
          2,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
        Object {
          "type": "return",
          "value": Promise {},
        },
        Object {
          "type": "return",
          "value": Promise {},
        },
        Object {
          "type": "return",
          "value": Promise {},
        },
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  loadNextPage={[Function]}
  onScroll={[Function]}
  onUpdateMatchedResults={[MockFunction]}
  spacerRef={
    Object {
      "current": null,
    }
  }
  status="SUCCESS"
  uniqField="id"
  wrapperRef={
    Object {
      "current": null,
    }
  }
/>
`;
